Expose `set_edited` and `is_edited` on `EditorInterface`

Naros opened the pull request:
Exposes several tools-based methods on `Object`.

The reason to expose these is simple.  If an editor plugin provides a `EditorResourcePreviewGenerator` implementation to create thumbnails for custom resources, there isn't a way for the plugin to have the `EditorResourcePreview` regenerate thumbnails when the resource is modified. In the engine, this is handled by calling `set_edited(true)`, which increments the edited version that is used by the hash algorithm to determine whether the cached thumbnail should be discarded and regenerated.

By exposing these three methods, addons can easily use the thumbnail generator framework and regenerate the thumbnails when the underlying resource is changed rather than requiring an editor restart.

Naros committed:
Expose `set_edited` and `is_edited` on EditorInterface

Naros commented on doc/classes/Object.xml:
```diff
@@ -971,6 +983,14 @@
+			<return type="void" />
+			<param index="0" name="edited" type="bool" />
+			<description>
+				Set whether the object has been edited based on the [param edited] value.
```
Updated with the suggestions, thanks.

AThousandShips commented on doc/classes/Object.xml:
```diff
@@ -971,6 +983,15 @@
+			<param index="0" name="edited" type="bool" />
+			<description>
+				Set whether the object has been edited based on the [param edited] value.
+				[b]Note:[/b] This is primarily used by the editor for [Resource] based objects to track their modified state. For example, changing a scene, a resource in the inspector, or a script will call this method with [code]true[/code]. Saving the scene, script, or resource reset the edited state by calling this method with [code]false[/code].
```
The "for example" wording needs some work I think, it's unclear what it means, I'd say "changing a scene" here sounds like switching tabs, maybe something like:
```suggestion
				[b]Note:[/b] This is primarily used by the editor for [Resource] based objects to track their modified state. For example, any changes to an open scene, a resource in the inspector, or an edited script will cause this method to be called with [code]true[/code]. Saving the scene, script, or resource resets the edited state by calling this method with [code]false[/code].
```

Mickeon reviewed (commented):
Starting thoughts, I'm not particularly fond of this. Mainly, these methods feel too "technical", "mechanical" to be needing to be exposed. The additional documentation does not help. Yes, these are used _by the engine_, but what can the _user_... use this for?
The above questioning is why we decided to "unexpose" Resource's `setup_local_to_scene` (https://github.com/godotengine/godot/pull/67082). The method is literally just an internal method exposed for little to no reason: The user could implement it on their own, and documenting its internal purpose was difficult. So we gotta be a bit careful to expose these willy-nilly.

A quick skim over this has me believe that it would be a better fit to expose this in **Resource**, instead.

And furthermore, these are purely editor-exclusive, and very situational. Too much to be in **Object**, personally.
This has me think there's got to be a better way to accomplish what one may need this for. Namely, exposing these methods through another Editor-exclusive class.

Naros commented:
> Starting thoughts, I'm not particularly fond of this. Mainly, these methods feel too "technical", "mechanical" to be needing to be exposed. The additional documentation does not help. Yes, these are used _by the engine_, but what can the _user_... use this for? 

Hi, @Mickeon. There are several reasons for exposing these methods.

**set_edited**:

It's common practice in when serializing sub-resources during a save operation to set the resource edited state to `false`. Without this method, any custom `ResourceFormatSaver` has no way to reset this state, and the engine will continue to view the resource as modified. This creates a divide in how C++ modules and GDExtension work for plug-ins, meaning that GDExtensions that have custom resources are incapable of handling this use case properly.

**is_edited**:

This one is chosen to be exposed mainly because if GDExtension needs a way to set the state to `true` or `false,` it may need a way in the future to query the edited state of the object.  So this is purely for convenience.

**get_edited_version**:

This was being exposed primarily to support `hash_edited_version_for_preview`; however, in hindsight, maybe it would make more sense to change this and expose `hash_edited_version_for_preview` instead so that again resources exposed in GDExtension that may want to disable previews, like `TileSet`, can do so.

> A quick skim over this has me believe that it would be a better fit to expose this in **Resource**, instead.
>
> And furthermore, these are purely editor-exclusive, and very situational. Too much to be in **Object**, personally. This has me think there's got to be a better way to accomplish what one may need this for. Namely, exposing these methods through another Editor-exclusive class.

For our purposes, exposing this on the `Resource` class would definitely work, given that's our prime use case; however, I wonder if placing these on `EditorInterface` would also be better for other non-resource types, such as `ConfigFile`?  

```cpp
void EditorInterface::set_object_edited(Object* p_object, bool p_edited);
bool EditorInterface::is_object_edited(Object* p_object);
```

As for the `get_edited_version` or `hash_edited_version_for_preview`, I think maybe the best play is to expose the latter on the `Resource` class with a GDVIRTUAL call that can be implemented by custom resources.

Whats your thoughts on that approach @Mickeon ?

Mickeon commented:
Sorry for not having much else to add. I am mostly looking at this from a general user perspective. Whether this is actually implemented should not be up to me.

> I wonder if placing these on `EditorInterface` would also be better for other non-resource types, such as `ConfigFile`?

As I hinted at, delegating this to another class may be a better choice. So, personally, yes.

Naros commented:
@AThousandShips so I decided on the following, in an attempt to get this into 4.4.

1. I removed exposing the `get_edited_version` from this PR. That is more directed toward rendering of thumbnails and I'd like to look at that work from a separate PR as I believe there may be other changes to align GDE and C++ modules in this area.
2. I removed exposing the `set_edited` and `is_edited` on `Object` and provided delegators from `EditorInterface` instead.

Let me know if you think anything else is needed here.

dsnopek reviewed (commented):
Thanks!

This makes sense to me. Having it on `EditorInterface` sounds good, because these are only for editor plugins. The docs seem pretty clear, although, I think they would be even better with @Mickeon's suggestions above

dsnopek reviewed (approved):
Looks good to me!

KoBeWi commented:
Any example plugin to test this?
I checked how it works with my plugin and setting or not setting edited does not seem to make any difference.

> This one is chosen to be exposed mainly because if GDExtension needs a way to set the state to true or false, it may need a way in the future to query the edited state of the object. So this is purely for convenience.

Sounds like theoretical use-case.

Naros commented:
In Orchestrator for example, we want to have a way to mark our custom `Resource` as edited when a user makes some UI modification. Right now we either need to implement this behavior with custom `is_edited` / `set_edited` methods on our resource or maintain this state elsewhere in the UI. 

In addition, there are several spots in the editor where it explicitly uses `Resource::is_edited`.  For example, in the `EditorFileSystem::_copy_file` method, the editor explicitly checks whether a resource is modified before it is copied, and if it is, it explicitly calls save on it to avoid changes being lost and discrepancies.

Another example is in `EditorNode::_find_and_save_resource`, it again checks whether the `Resource::is_edited`, and only triggers the `ResourceSaver::save` call if the resource is flagged as edited. The same also applies in `EditorNode::_save_external_resources`.

A custom resource that implements its own edited tracking can't make use of any of this behavior at all.

So it's not only just a matter of convenience for tracking the state within one's own plugin, but it's also about making sure our custom resources work just like any other Godot resource does, @KoBeWi .

AThousandShips commented on editor/editor_interface.cpp:
```diff
@@ -659,6 +659,16 @@ void EditorInterface::reload_scene_from_path(const String &scene_path) {
+}
+
+bool EditorInterface::is_object_edited(Object *p_object) const {
+	ERR_FAIL_COND_V_MSG(!p_object, false, "Cannot check edit status on a null object");
```
```suggestion
	ERR_FAIL_NULL_V_MSG(p_object, false, "Cannot check edit status on a null object.");
```

KoBeWi reviewed (approved):
Still not convinced about `is_object_edited()`, but otherwise looks fine.
